posthog/queries/person_query.py

```diff
@@ -94,10 +94,21 @@ def get_query(self, prepend: str = "", paginate: bool = False) -> Tuple[str, Dic
             FROM person
             WHERE team_id = %(team_id)s
             AND id IN (
-                SELECT id FROM person
-                {cohort_query}
-                WHERE team_id = %(team_id)s
-                {person_filters}
+                SELECT id FROM (
+                    SELECT *
+                    FROM person
+                    {cohort_query}
+                    WHERE team_id = %(team_id)s
+                    ORDER BY version DESC
+                    LIMIT 1 BY id
+                    -- NOTE: by default asterisk doesn't include materialized
+                    -- columns. I don't know how optimised this is e.g. if the
+                    -- stream returned by this subquery will include literally
+                    -- everything of if it will be optimised to only include
+                    -- those columns that are actually used in the outer query.
+                    SETTINGS asterisk_include_materialized_columns=1
+                ) AS person
+                WHERE True {person_filters}
             )
```

Comment:
Continuing from slack thread, what's incorrect about filter on everything -> get max version of properties -> filter again? If i'm understanding correctly, right now the query will filter on everything which gives the relevant IDs of _potentially correct_ persons and then gets the latest properties on those persons to filter again to get the exact set of right persons

The changes here seems correct but the ordering may cause a regression to low performance.

context: https://posthog.slack.com/archives/C046D7C654H/p1668008520458769

Reply:
Hmm the max is an argMax so `argMax(property, version)` would give "another thing" and then get filtered by the second layer is how I'm understanding it

Reply:
So

```
SELECT argMax(properties, version) as person_props
// gets a row where id = 0 and property = "another thing"
...
FROM persons where id IN
(
     SELECT id FROM persons where properties = "one thing"
     // returns id = 0
)
...

GROUP BY id
HAVING max(is_deleted) = 0
AND person_props = "one thing"
// filters away "another thing"
```
return no matches

Reply:
But: 

```SELECT argMax(properties, version) AS person_props
// gets a row WHERE id = 0 AND property != "another thing"
...
FROM persons WHERE id IN
(
     SELECT id FROM persons WHERE properties != "another thing"
     // returns id = 0
)
...

GROUP BY id
HAVING max(is_deleted) = 0
AND person_props != "another thing"
// filters away "another thing"
```

returns version = 0 but it shouldn't

Reply:
i.e you're returning the latest version for which the condition is true.

Reply:
(I could be missing something, it's late)

Reply:
Huh, wait does the return happen? Inner query returns id=0, outer query will match that ID and then return properties based on latest version (so "another thing" again) but person_props filter at the end will filter it away so the return should be no matches

Reply:
Even with the `properties = "one thing"` is that not wrong? It will return a list where the condition is true, then you test again and it's still true.

Reply:
it returns a list of IDs where it's true but the properties themselves aren't there. The ids are then used (without consideration to the properties that it was filtered on)

if i'm understanding how his query ends up working is that it doesn't exactly matter what the inner query returns. 
1. It does an initial filtering on the complete set of persons to reduce the considered set down to something reasonable. 
2. The set here contains both possible matches and faulty matches on historic person properties. We _only return the ids_ to the outer query so the outer query simply uses those IDs and retrieves the latest properties associated to those ids based on the max version number. 
3. This gives you the up to date properties on the _possible matches_ from the inner query. Then, you filter it again to get rid of the duds

Reply:
The id is not unique btw.

Reply:
I think that's the confusion?

Reply:
I think the confusion is thinking that ClickHouse is a relational database?

Reply:
More context here: https://posthog.slack.com/archives/C046D7C654H/p1668008520458769